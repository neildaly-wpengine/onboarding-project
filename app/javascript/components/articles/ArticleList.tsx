import {
  Button,
  createStyles,
  Fab,
  Grid,
  makeStyles,
  Snackbar,
  Theme,
} from "@material-ui/core";
import Fade from "@material-ui/core/Fade";
import AddIcon from "@material-ui/icons/Add";
import React, { useEffect, useState } from "react";
import { Link, useHistory } from "react-router-dom";
import {
  Article,
  ArticleListLocationState,
  AuthStoreProps,
  ConsumerProps,
  DeleteResponse,
} from "../../common/types";
import CollapsibleAlert from "../alert/CollapsibleAlert";
import ArticleHighlight from "./ArticleHighlight";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    grid: { margin: 0, width: "100%" },
    link: {
      textDecoration: "none",
      color: "#fff",
      lineHeight: 0,
    },
    fab: {
      margin: 0,
      top: "auto",
      right: 20,
      bottom: 20,
      left: "auto",
      position: "fixed",
    },
    snackbar: {
      [theme.breakpoints.down("xs")]: {
        bottom: 90,
      },
    },
  })
);

const ArticleList: React.FC<
  ConsumerProps & AuthStoreProps & ArticleListLocationState
> = ({ consumer, authStore, location }) => {
  const [articles, setArticles] = useState<Article[]>();
  const [alertMessage, setAlertMessage] = useState<string>("");
  const [deletedArticle, setDeletedArticle] = useState<string>("");
  const classes = useStyles();
  const history = useHistory();

  useEffect(() => {
    let mounted: boolean = true;
    const fetchAllArticles = async () => {
      const articleData = await consumer.getAllArticles();
      if (mounted) {
        setArticles(articleData);
      }
    };
    fetchAllArticles();

    // Handle article creation flash
    const handleAlertFlash = (): void => {
      if (mounted) {
        if (location.state !== undefined && location.state.message) {
          setAlertMessage(location.state.message);
          const state = location.state;
          delete state.message;
          history.replace({ ...history.location, state });
        }
      }
    };
    handleAlertFlash();

    return () => {
      mounted = false;
    };
  }, []);

  if (articles === undefined) {
    return null;
  }

  const closeAlert = (): void => {
    setAlertMessage("");
  };

  const handleArticleDeletion = (articleID: string) => {
    consumer.deleteArticle(articleID).then((response: DeleteResponse) => {
      if (response.data.discardedAt) {
        setArticles(
          articles.filter((article: Article) => {
            return article.id !== articleID;
          })
        );
        setDeletedArticle(articleID);
      }
    });
  };

  const closeSnackbar = () => {
    setDeletedArticle("");
  };

  const undoDeletion = async () => {
    const recoveredArticle: Article = await consumer.recoverArticle(
      deletedArticle
    );
    setDeletedArticle("");
    setAlertMessage("Article Recovered!");
    setArticles([recoveredArticle].concat(articles));
  };

  const articlesList = articles.map((article: Article) => {
    return (
      <ArticleHighlight
        key={article.id}
        article={article}
        auth={{
          authenticated: authStore.authenticated,
          currentUserID: authStore.user.id,
        }}
        notifyDelete={handleArticleDeletion}
      />
    );
  });

  const createArticleMarkup: JSX.Element = authStore.authenticated ? (
    <Fab
      color="secondary"
      aria-label="add"
      data-testid="create-fab"
      className={classes.fab}
    >
      <Link to="/create" className={classes.link}>
        <AddIcon />
      </Link>
    </Fab>
  ) : (
    <></>
  );

  return (
    <React.Fragment>
      <CollapsibleAlert
        message={alertMessage}
        severity="success"
        showAlert={alertMessage !== ""}
        closeAlert={closeAlert}
      />
      <Grid container data-testid="resolved" style={{ marginTop: 25 }}>
        <Grid item xs={12}>
          <Fade in={true}>
            <Grid container justify="center" className={classes.grid}>
              {articlesList}
            </Grid>
          </Fade>
        </Grid>
      </Grid>
      {createArticleMarkup}
      <Snackbar
        open={deletedArticle !== ""}
        autoHideDuration={6000}
        onClose={closeSnackbar}
        message="Article Deleted"
        action={
          <Button color="inherit" size="small" onClick={undoDeletion}>
            Undo
          </Button>
        }
        className={classes.snackbar}
      />
    </React.Fragment>
  );
};

export default ArticleList;
