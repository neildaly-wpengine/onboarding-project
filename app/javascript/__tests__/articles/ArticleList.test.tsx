import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import { cleanup, render, RenderResult, waitFor } from "@testing-library/react";
import axios from "axios";
import * as React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import ArticleList from "../../components/articles/ArticleList";
import {
  consumer,
  jsonAPIAllArticleResponseData,
} from "./__helpers__/test-data";

const renderArticleList = (): RenderResult => {
  return render(
    <Router>
      <ArticleList consumer={consumer} />
    </Router>
  );
};

afterEach(() => {
  cleanup;
});

jest.mock("axios");

describe("<ArticleList />", () => {
  describe("Content", () => {
    it("renders correctly", async () => {
      const expectedResponse = {
        data: jsonAPIAllArticleResponseData,
      };

      (axios.get as jest.Mock).mockImplementationOnce(() =>
        Promise.resolve(expectedResponse)
      );

      const { container, getByTestId } = renderArticleList();
      expect(getByTestId("loading")).toHaveAttribute("role", "progressbar");

      const resolvedData = await waitFor(() => getByTestId("resolved"));

      expect(container).toMatchSnapshot();
      expect(axios.get).toHaveBeenCalledWith("/api/v1/articles");
      expect(resolvedData).toBeInTheDocument();
      expect(resolvedData).not.toBeEmptyDOMElement();
      expect(axios.get).toBeCalledTimes(1);
    });
  });
});
