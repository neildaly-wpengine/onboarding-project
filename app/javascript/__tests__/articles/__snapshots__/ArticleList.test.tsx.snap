// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleList /> Content renders correctly when authenticated 1`] = `
<div>
  <div
    class="MuiCollapse-container MuiCollapse-hidden"
    style="min-height: 0px;"
  >
    <div
      class="MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner"
      >
        <div
          class="MuiPaper-root MuiAlert-root MuiAlert-standardSuccess makeStyles-alert-14 MuiPaper-elevation0"
          role="alert"
        >
          <div
            class="MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message"
          />
          <div
            class="MuiAlert-action"
          >
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container"
    data-testid="resolved"
    style="opacity: 1; margin-top: 25px; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiGrid-root makeStyles-grid-10 MuiGrid-container MuiGrid-justify-xs-center"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-root-15 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardHeader-root"
            data-testid="article-card-header"
          >
            <div
              class="MuiCardHeader-avatar"
            >
              <div
                aria-label="user"
                class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-17 MuiAvatar-colorDefault"
              >
                HS
              </div>
            </div>
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
              >
                Harry Smith
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                2020-06-25T10:07:42.357Z
              </span>
            </div>
            <div
              class="MuiCardHeader-action"
            >
              <button
                aria-label="settings"
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiCardActionArea-root"
            tabindex="0"
            type="button"
          >
            <img
              alt="blog-image"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
              height="200"
              src="https://picsum.photos/400/200?image=1"
              title="Computer Games Development"
            />
            <div
              class="MuiCardContent-root"
            >
              <h2
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
                data-testid="article-card-title"
              >
                Computer Games Development
              </h2>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                data-testid="article-card-content"
              >
                Lorem ipsum dolor sit 
                ...
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiCardActions-root makeStyles-cardActions-16 MuiCardActions-spacing"
          >
            <button
              aria-label="add to favorites"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
              data-testid="article-card-button"
              href="/articles/1"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View More
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    aria-label="add"
    class="MuiButtonBase-root MuiFab-root makeStyles-fab-12 MuiFab-secondary"
    data-testid="create-fab"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label"
    >
      <a
        class="makeStyles-link-11"
        href="/create"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </a>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`<ArticleList /> Content renders correctly when not authenticated 1`] = `
<div>
  <div
    class="MuiCollapse-container MuiCollapse-hidden"
    style="min-height: 0px;"
  >
    <div
      class="MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner"
      >
        <div
          class="MuiPaper-root MuiAlert-root MuiAlert-standardSuccess makeStyles-alert-5 MuiPaper-elevation0"
          role="alert"
        >
          <div
            class="MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message"
          />
          <div
            class="MuiAlert-action"
          >
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container"
    data-testid="resolved"
    style="opacity: 1; margin-top: 25px; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiGrid-root makeStyles-grid-1 MuiGrid-container MuiGrid-justify-xs-center"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-root-6 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardHeader-root"
            data-testid="article-card-header"
          >
            <div
              class="MuiCardHeader-avatar"
            >
              <div
                aria-label="user"
                class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-8 MuiAvatar-colorDefault"
              >
                HS
              </div>
            </div>
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
              >
                Harry Smith
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                2020-06-25T10:07:42.357Z
              </span>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiCardActionArea-root"
            tabindex="0"
            type="button"
          >
            <img
              alt="blog-image"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
              height="200"
              src="https://picsum.photos/400/200?image=1"
              title="Computer Games Development"
            />
            <div
              class="MuiCardContent-root"
            >
              <h2
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
                data-testid="article-card-title"
              >
                Computer Games Development
              </h2>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                data-testid="article-card-content"
              >
                Lorem ipsum dolor sit 
                ...
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiCardActions-root makeStyles-cardActions-7 MuiCardActions-spacing"
          >
            <button
              aria-label="add to favorites"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
              data-testid="article-card-button"
              href="/articles/1"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View More
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
